### PR TITLE
Test/ddp

### PR DESCRIPTION
**Pull request by bobbycxy**
I am adding Distributed Data Parallel (DDP) capabilities to the current code base. Some minor surgery was required to ensure we could perform DDP. At each line of code that was changed or newly added, comments were added so the reviewer may understand what took place. In the following, only the key changes will be highlighted. As for other changes, please refer to the comments.

### Key Changes:
**A. 'base_trainer.py'**
1. Created a new class `CustomDataset` that would wrap around the existing `dataloader` so that it can be fed into Pytorch's `DataLoader` for DDP.
2. Used the DDP function to wrap around the `self.model`, and assigned gpu_id and batch_size into the BaseTrainer's self attributes.
3. Divided the self.gradient_accumulation_steps by torch.cuda.device_count() to maximise throughput
4. Where printing log results took place, I added a condition to print from GPU:0 instead.
5. the _run_steps function takes in a new argument called 'epoch' to ensure Pytorch's `DataLoader` for DDP will be shuffled ([discussion](https://discuss.pytorch.org/t/why-is-sampler-set-epoch-epoch-needed-for-distributedsampler/149672/2)). 

**B. build_trainers.py**
1. Added a function called ddp_steup. This is used to initialise the DDP process. 

**C. dataloader.py**
1. Just added a 'get_batch(self, split = 'train')' method into the BaseDataLoader class. This is used in unison with `CustomDataset`.

**D. utils.py**
1. Added a function called aggregate_value that, as name suggest, aggregates the values across tensors. More specifically, it averages. Will it ever have to sum? hmm.. Currently, it's used in the base_trainer and metrics .py scripts.

**E. metrics.py**
1. It turns out that the functions used here also calculate the metrics across tensors. Hence, the aggregate value needs to be used here. 

**F. eval_wrapper.py**
1. Wherever a model's method gets called, the use of DDP will mean that we have to add '.module' between the method and the model ([discussion](https://discuss.pytorch.org/t/access-to-attributes-of-model-wrapped-in-ddp/130572/2))

**G. train.py**
1. A refactor of the main is required to ensure the main() continues to allow for hydra's insertion of cfg into the main(). 
2. Key addition is the destroy_process_group() which is necessary for DDP.

### CODE REVIEW

File: trainers/base_trainer.py [hunk 1]
```diff
@@ -43,18 +79,19 @@ def __init__(
         self.cfg = cfg
         self.gradient_accumulation_steps = cfg["trainer"]["training"][
             "gradient_accumulation_steps"
-        ]
+        ] // torch.cuda.device_count() ## divide by number of GPUs to maximise throughput
```
Comment (DylanASHillier) [inline trainers/base_trainer.py:82]: what happens when we have say an odd number of grad accumulation steps and even number of gpus (or more likely 3 gpus and say 20 gpu acc. steps)

Comment (DylanASHillier) [reply trainers/base_trainer.py:82]: tbh not sure what the best idea here is ... it may be that we want to assert that it is divisible by 3 or sth. not sure

File: trainers/dataloader.py [hunk 2]
```diff
@@ -135,6 +135,17 @@ def process(example):
             # if we fail, destroy the file
             os.removedirs(self.tokenized_data_path)
             raise SystemExit from exc
+        
+    def get_data(self, split="train"):
+        """
+        Get the data
+        """
```
Comment (DylanASHillier) [inline trainers/dataloader.py:142]: this is a good method to have keep this

File: trainers/base_trainer.py [hunk 3]
```diff
@@ -16,6 +16,39 @@
 )
 from trainers.evaluator import train_eval
 
+import numpy as np
+from itertools import islice
+from torch.nn.parallel import DistributedDataParallel as DDP 
+from torch.utils.data.distributed import DistributedSampler
+from trainers.utils import aggregate_value
+
+class CustomDataset:
```
Comment (DylanASHillier) [inline trainers/base_trainer.py:25]: Pointless having this here. Make our dataclass subclass the torch dataloader. Then replace the get_batch  method with this get_item method

File: trainers/base_trainer.py [hunk 4]
```diff
@@ -109,20 +158,28 @@ def estimate_performance(self, eval_iters=None):
         for split in ["train", "val"]:
             losses = torch.zeros(eval_iters)
             perplexities = torch.zeros(eval_iters)
-            for i in range(eval_iters):
+            
+            ## init Pytorch's DataLoader
+            dataloader = self._prepare_dataloader(CustomDataset(self.dataloader.get_data(split), self.cfg.model["context_window"], self.cfg.general.device))
```
Comment (DylanASHillier) [inline trainers/base_trainer.py:163]: prepare the dataloader once for train and eval and store these as attributes in the trainer. then just iterate over them. since the data doesn't change, there is no point of doing it like this

Comment (DylanASHillier) [reply trainers/base_trainer.py:163]: which will be anyways what you do if you our dataloader inherit from the torch dataset

File: trainers/base_trainer.py (hunk 3 above)
Comment (DylanASHillier) [reply trainers/base_trainer.py:25]: * make our dataloader subclass/inherit the torch dataset

**Review by DylanASHillier — COMMENTED** [review]
Mostly great. Change our dataloader to inherit from the torch dataset to avoid the faff of initiating a torch dataloader separately. Make sure to update the other ones by replacing their get_batch functions

**Comment by DylanASHillier** [comment]
not sure what we should do about the prints. I think having a global variable that just checks the rank is 0 maybe that you can then reference. Sort of similar to here: https://stackoverflow.com/questions/66261729/pytorch-lightning-duplicates-main-script-in-ddp-mode

**Comment by DylanASHillier** [comment]
To summarize (think you can tick off too?)
- [ ] Refactor the customdataset/ our original dataloader into one piece of code by replacing the get_batch with the get_data of customdataset
- [ ] propogate this to the other dataloaders that inherit from this
- [ ] don't reinitialize the dataloader every epoch, just use a single one for train and eval each
- [ ] find a good solution for the prints

**Comment by bobbycxy** [comment]
> To summarize (think you can tick off too?)
> 
> * [x]  Refactor the customdataset/ our original dataloader into one piece of code by replacing the get_batch with the get_data of customdataset
> * [x]  propogate this to the other dataloaders that inherit from this
> * [x]  don't reinitialize the dataloader every epoch, just use a single one for train and eval each
> * [x]  find a good solution for the prints

Achieved changes 1, 2 and 3 by
1. Creating a shared get_data() method in dataloader.py that tailors to the data grabbing conditions used by ByteDataLoader and ConversationalDataLoader. Just so you know, the 'get_data()' method is found in the parent class of BaseDataLoader. 
2. Leveraging a shared __getitem__ and shared __len__ method across StandardDataLoader and ByteDataLoader. A custom __getitem__ and __len__ were written for ConversationDataLoader considering, and a custom __getitem__ was written for NextTokenMLMDataLoader.
3. In addition, just for the NextTokenMLMDataLoader, I preassigned a default 0.15 `masking_pct` value unless stated in the config dictionary.
4. Rewrote the _prepare_dataloader() by including the set_data(split) **within** the _prepare_dataloader(). In addition, the _prepare_dataloader **will use a cached train or val dataloader if one exists**. If none, then it will be created from scratch. These train and val dataloader are assigned to the base_trainer attributes as self.tran_val_dataloaders:dict

Achieved change 4 by
1. Creating `init_print_override()` and `restore_print_override(original_print)` that basically adjusts the usual `print` function to now print when os.getenv('GLOBAL_RANK') == '0'.
2. Editted the ddp_main() function in train.py to 'try' and should there be an error, the 'finally' will destroy the process group and always restore the print override. 

Let me know if the edits are okay. If all good, I will do some tidying up of the code. Namely, removing get_batch (left it there so it's easier to review the changes)

**Comment by bobbycxy** [comment]
[Implement DDP](https://trello.com/c/cvwiMsAk/2-implement-ddp)

**Comment by DylanASHillier** [comment]
just merge lah wait for what

File: evals/eval_wrapper.py [hunk 5]
```diff
@@ -28,7 +28,7 @@ def loglikelihood(self, prefixes, continuations) -> list[float]:
                 for prefix_batch, cont_batch in zip(
                     batch(prefixes, 32), batch(continuations, 32)
                 ):
-                    ll = self.model_shell.loglikelihood(prefix_batch, cont_batch)
+                    ll = self.model_shell.loglikelihood(prefix_batch, cont_batch) ## added the 'module' attribute (https://discuss.pytorch.org/t/access-to-attributes-of-model-wrapped-in-ddp/130572/2)
```
Comment (DylanASHillier) [inline evals/eval_wrapper.py:31]: bro why is this still not removed :(

Comment (DylanASHillier) [reply evals/eval_wrapper.py:31]: oh its just the comment. still remove the comment lol

File: evals/metrics.py [hunk 6]
```diff
@@ -15,7 +16,9 @@ def accuracy_metric(confidences):
         accuracy: float of accuracy
     """
     _, predicted = torch.max(confidences, 1)
-    return (predicted == 0).float().mean()
+    ## aggregate the tensor values
+    mean_accuracy = aggregate_value((predicted == 0).float().mean())
+    return mean_accuracy#(predicted == 0).float().mean()
```
Comment (DylanASHillier) [inline evals/metrics.py:21]: could probs be done in one line...

File: evals/metrics.py [hunk 7]
```diff
@@ -29,7 +32,9 @@ def path_confidence(confidences):
     """
     softmaxed = torch.nn.functional.softmax(confidences, dim=-1)
     softmaxed = softmaxed[:, 0]
-    return softmaxed.mean()
+    ## aggregate the tensor values
+    mean_confidence = aggregate_value(softmaxed.mean())
+    return mean_confidence
```
Comment (DylanASHillier) [inline evals/metrics.py:37]: likewise

File: trainers/base_trainer.py [hunk 8]
```diff
@@ -33,30 +39,36 @@ def __init__(
         optimizer,
         dataloader: train_dataloader.BaseDataloader,
         loss_fn,
+        gpu_id, ## new
         lr_scheduler=None,
         dropout_scheduler=None,
     ) -> None:
         self.model = model
+        self.model = DDP(self.model, device_ids=[gpu_id]) ## new
```
Comment (DylanASHillier) [inline trainers/base_trainer.py:47]: wonder what the effect of having this as a seperate attribute of trainer would be i.e. self.ddp_model = ... -- cos then we can still pass the underlying model to other methods, and just change how we call the forward pass to call self.ddp_model

File: trainers/base_trainer.py [hunk 9]
```diff
@@ -197,10 +255,10 @@ def run_profile(self):
         ) as prof:
             for i in range(10):
                 if i <= 3:
-                    self._run_step()
+                    self._run_step(i) ## set the 'epoch' to ensure shuffle
```
Comment (DylanASHillier) [inline trainers/base_trainer.py:258]: btw is this still necessary?

File: trainers/dataloader.py [hunk 10]
```diff
@@ -135,6 +138,49 @@ def process(example):
...
+            mode="r",
+        )
+
+        ## additional steps for loading BytePooling and Conversational dataloaders
+        if hasattr(self, 'loading_shapes') and self.loading_shapes[split] is None:
+            if isinstance(self, BytePoolingDataloader):
+                self.loading_shapes[split] = (len(data)// self.model_cfg.embedder.byte_context_window, self.model_cfg.embedder.byte_context_window)
+
+                ## reset the data
+                data = None
```
Comment (DylanASHillier) [inline trainers/dataloader.py:159]: we probs want to rework this in the future... these should be delegated to the subclassing dataloaders... add a todo for that please

**Review by DylanASHillier — APPROVED** [review]
Approved again, again a few changes to make first but quite minor

File: trainers/dataloader.py (hunk 10 above)
Comment (DylanASHillier) [reply trainers/dataloader.py:159]: i.e. instead of if else the load data should be handled by each subclass of the dataloader

Comment (DylanASHillier) [reply trainers/dataloader.py:159]: i actually hate isinstance(self, ...) - that implies something has gone horribly wrong imo

Comment (DylanASHillier) [reply trainers/dataloader.py:159]: https://stackoverflow.com/questions/53104810/check-if-self-is-instance-of-subclass-in-python

File: trainers/base_trainer.py (hunk 9 above)
Comment (bobbycxy) [reply trainers/base_trainer.py:258]: Yes. I implemented it without it previously and I printed the X and y. They kept returning the same sets over each iteration. In other words, gradient_accumulation_step 0 returns the same across all iterations. That led to a smooth loss value. 

Revisiting the Pytorch DDP implementation and following this [thread](https://discuss.pytorch.org/t/why-is-sampler-set-epoch-epoch-needed-for-distributedsampler/149672/2), setting the epoch will help to shuffle the data. This means the gradient_accumulation_step 0 for iteration 0 and iteration 1 will no longer be the same.

File: trainers/dataloader.py (hunk 10 above)
Comment (bobbycxy) [reply trainers/dataloader.py:159]: can, i'll add it so. And you're right, I think subclass is better. Initially compiled it so that the subclasses will look neater.

File: trainers/dataloader.py [hunk 11]
```diff
@@ -231,6 +200,28 @@ def _write_tokenized_data(self, tokenized):
...
+
+            ## reset the data
+            data = None
+
+        ## re-load the data with loading shapes
+        data = np.memmap(
+            os.path.join(self.tokenized_data_path, f"{split}.bin"),
+            dtype=np.uint16,
+            mode="r",
+            shape=self.loading_shapes[split],
```
Comment (DylanASHillier) [inline trainers/dataloader.py:221]: I mean this could just be a subclassed function _get_loading_shape() rather than subclassing full func. but this is good for now lol

Comment (DylanASHillier) [reply trainers/dataloader.py:221]: and tbf this could change more for different ones so happy to keep like this
